llama-index-integrations/readers/llama-index-readers-oracleai/llama_index/readers/oracleai/base.py

```diff
@@ -0,0 +1,457 @@
+                    )
+
+                    cursor.execute(sql)
+                    for row in cursor:
+                        metadata = {}
+
+                        if row is None:
+                            doc_id = OracleDocReader.generate_object_id(
+                                self.conn.username
+                                + "$"
+                                + self.owner
+                                + "$"
+                                + self.tablename
+                                + "$"
+                                + self.colname
+                            )
+                            metadata["_oid"] = doc_id
+                            results.append(Document(text="", metadata=metadata))
+                        else:
+                            if row[0] is not None:
+                                data = str(row[0])
+                                if data.startswith(("<!DOCTYPE html", "<HTML>")):
+                                    p = ParseOracleDocMetadata()
+                                    p.feed(data)
+                                    metadata = p.get_metadata()
+
+                            doc_id = OracleDocReader.generate_object_id(
+                                self.conn.username
+                                + "$"
+                                + self.owner
+                                + "$"
+                                + self.tablename
+                                + "$"
+                                + self.colname
+                                + "$"
+                                + str(row[2])
+                            )
+                            metadata["_oid"] = doc_id
+                            metadata["_rowid"] = row[2]
+
+                            # process projected metadata cols
+                            if self.mdata_cols is not None:
+                                ncols = len(self.mdata_cols)
+
+                            for i in range(ncols):
+                                if i == 0:
+                                    metadata["_rowid"] = row[i + 2]
+                                else:
+                                    metadata[self.mdata_cols[i]] = row[i + 2]
+
+                            if row[1] is None:
+                                results.append(Document(text="", metadata=metadata))
+                            else:
+                                results.append(
+                                    Document(text=str(row[1]), metadata=metadata)
+                                )
+                except Exception as ex:
+                    print(f"An exception occurred :: {ex}")
+                    traceback.print_exc()
+                    cursor.close()
+                    raise
+
+            return results
+        except Exception as ex:
+            print(f"An exception occurred :: {ex}")
+            traceback.print_exc()
+            raise
+
+
+import logging
+from typing import (
+    Any,
```

Comment:
We should put all imports at the top imo

Reply:
moved.

Reply:
thanks